### PR TITLE
CUDA: Fix #5820, adding atomic nanmin / nanmax

### LINKED ISSUE #5820: Add back support for atomics that always return the non-NaN values
## Feature request

From @uchytilc in #5556:
> I saw this change in the change log and was wondering if there would be a way to revert the change via an option or something to allow atomic min/max to return the non-Nan value always. I have found this a very useful feature. For example it prevents the need to initialize different values when trying to find a max/min reduction (i.e. -inf vs +inf).

### PR DESCRIPTION
**Pull request by gmarkall**
This continues on from PR #5941, to fix #5820 by adding implementations of atomic `nanmin` and `nanmax` for CUDA.

It turned out that the `max` and `min` atomic functions also failed to return the old value when it was replaced, so a fix for this is included.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for carrying on with this @gmarkall. Following on from @sklam's review, https://github.com/numba/numba/pull/5941#discussion_r447589060 needs checking/doing if not done.

File: numba/cuda/cudadrv/nvvm.py [hunk 1]
```diff
@@ -423,25 +423,23 @@ def get(self):
...
     ; Check if either value is a NaN and return *ptr early if so
-    %valnan = fcmp uno {T} %val, %ptrval
+    %valnan = fcmp uno {T} %val, %{PTR_OR_VAL}val
     br i1 %valnan, label %done, label %gt_check
 
 gt_check:
     %dold = phi {T} [ %ptrval, %entry ], [ %dcas, %attempt ]
     ; Continue attempts if dold > val
-    %lt = fcmp nnan ogt {T} %dold, %val
+    %lt = fcmp {OP} {T} %dold, %val
```
Comment (stuartarchibald) [inline numba/cuda/cudadrv/nvvm.py:442]:
```suggestion
    %gt = fcmp {OP} {T} %dold, %val
```
should probably be called `gt`? But moot/ call it `cmp`? if the templates are unified.

File: numba/cuda/cudadrv/nvvm.py [hunk 2]
```diff
@@ -400,17 +400,17 @@ def get(self):
 
 
 ir_numba_atomic_max = """
-define internal {T} @___numba_atomic_{T}_max({T}* %ptr, {T} %val) alwaysinline {{
+define internal {T} @___numba_atomic_{T}_{NAN}max({T}* %ptr, {T} %val) alwaysinline {{
```
Comment (stuartarchibald) [inline numba/cuda/cudadrv/nvvm.py:403]: I think the two templates for max/min can be unified following the latest changes?

**Comment by esc** [comment]
BFID: `numba_smoketest_cuda_71`

**Comment by sklam** [comment]
Re-run buildfarm since https://github.com/numba/numba/pull/6016 merged to addresses the failed test (https://github.com/numba/numba/pull/6020#pullrequestreview-456003100)

BFID: numba_smoketest_cuda_72

**Pull request merged**